README.md
# Mastermind v2

<p align="center">
  <img src="https://miro.medium.com/max/700/0*4xohKyVpcqZPmT9y.png">
</p>

# Mastermind

This repository contains a Mastermind API written in Express and Typescript. Users can play the game with an included command line application.

## Table of Contents

- [Dependencies](#dependencies)
  - [Node](#node)
  - [NPM or YARN](#npm-or-yarn)
  - [Git](#git)
  - [Installation](#installation)
- [How to Play](#how-to-play)
- [Data Persistence](#data-persistence)
- [Development](#development)
  - [API Docs](#api-docs)
  - [GraphQL Server](#graphql-server)
  - [Testing](#testing)
  - [Continuous Integration](#continuous-integration)
  - [Logging](#logging)
- [Authors](#authors)

## Dependencies

### Node

- Node 14.0 or above is recommended to run this application.
- [Download Node](https://nodejs.org/en/download/)

### NPM or Yarn

- This is required to install all the node packages.
- NPM comes with node
- Yarn can be installed with the following command: `npm install -g yarn`

### Git

- [Download Git](https://git-scm.com/downloads)

### Installation

- On your terminal, clone the repository with Git:

`git clone https://github.com/tuvo1106/mastermind_v2`

- To install dependencies, use::

`npm install`

- To start the Express application, use:

`npm run start`

- To play the game, use:

`npm run cli`

## How to Play

In Mastermind, the player (aka the Codebreaker) has to guess a random sequence of four numbers generated by the Codemaker, which can either be the computer or another player. The Codebreaker wins when they guess both the number and order correctly for all four numbers within a given amount of turns. The breakdown of the different difficulties is shown below:

| Difficulty | Number Range | Number of Turns |
| :--------: | :----------: | :-------------: |
|    Easy    |    0 - 7     |       12        |
|   Normal   |    0 - 7     |       10        |
|    Hard    |    0 - 7     |        8        |

Each guess is made by typing in 4 consecutive numbers in the terminal.
After each guess, the Codebreaker will get feedback on the previous
guess. The feedback is represented in the following symbols:

```
X = Correct number, correct location
O = Correct number, incorrect location
* = Incorrect number, incorrect location
```

For example, if the correct sequence is [1, 9, 0, 2] and the player guesses
[1, 2, 3, 4], the feedback that is given will be:

`X O * *`

- The (1) `X` represents the 1 being the right number in the right position.
- The (1) `O` represents the 2 being the right number, but in the wrong positon.
- The (2) `*` represents 3 and 4 both being incorrect numbers and therefore, not
  belonging on the board.

In another example, if the correct sequence is [3, 3, 6, 9] and the player
guesses [1, 3, 6, 9], the feedback that is given will be:

`X X X *`

- The (3) `X` represents 3, 6 and 9 being the right numbers in the right
  positions.
- The (1) `*` represents the remaining unmatched number 3 being compared with
  the only unmatched guess which is 1, therefore resulting in a mismatch of
  number and position.

## Data Persistence

By default, the app stores all game data in memory. However, users can choose to store data with MongoDB by using Docker containers.

This requires an installation of a Docker client on your local operating system.

- [Download Docker](https://www.docker.com/products/docker-desktop)

- To start the MongoDB and Express app with Docker, use:

`npm run docker`

In the background, this runs `docker compose build && docker compose up` to start up the containers.

Express will be running on port `3000` and MongoDB will be running on port `27017`.

## Development

### Environmental Variables

You can set the following environmental variables to customize the app.

- `NODE_ENV` ("development" or "test" or "production", defaulted to "development")
- `DB` ("local" or "mongo", defaulted to "local")
- `MONGO_PORT` (defaulted to 27017)
- `PORT` (defaulted to 3000)

### API Docs

API docs can be found at `http://localhost:3000/api-docs` when the API is running.

![swagger](/src/assets/swagger.png)

### GraphQL Server

Users can also interact with the API using a GraphQL playground at `http://localhost:3000/graphql` when the API is running.

![graph-ql](/src/assets/graphql.png)

### Testing

To run tests, use:

`npm run test`

There are currently `55` total tests in this project with `84.9%` test coverage. (5/6/21)

```
-----------------------------------|---------|----------|---------|---------|----------------------------------------------------
File                               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-----------------------------------|---------|----------|---------|---------|----------------------------------------------------
All files                          |   82.99 |    65.03 |   81.69 |    84.9 |
 src                               |   97.67 |      100 |   66.67 |     100 |
  app.ts                           |   97.67 |      100 |   66.67 |     100 |
 src/application/enums             |     100 |      100 |     100 |     100 |
  appStatus.ts                     |     100 |      100 |     100 |     100 |
  gameStatus.ts                    |     100 |      100 |     100 |     100 |
 src/application/errors            |     100 |       50 |     100 |     100 |
  bad-request-error.ts             |     100 |       50 |     100 |     100 | 7
  custom-error.interface.ts        |     100 |       50 |     100 |     100 | 5
  not-found-error.ts               |     100 |       50 |     100 |     100 | 7
 src/application/factories         |     100 |       50 |     100 |     100 |
  game-factory.ts                  |     100 |      100 |     100 |     100 |
  user-factory.ts                  |     100 |       50 |     100 |     100 | 33
 src/application/middlewares       |    87.5 |       50 |     100 |   85.71 |
  error-handler.ts                 |    87.5 |       50 |     100 |   85.71 | 17
 src/application/utils             |     100 |       50 |     100 |     100 |
  createDirIfNotExists.ts          |     100 |       50 |     100 |     100 | 4
 src/domain/game                   |   97.02 |     89.8 |   96.88 |    96.6 |
  game-controller.ts               |     100 |      100 |     100 |     100 |
  game-params-validator-service.ts |      95 |    91.67 |     100 |      95 | 15
  game-service.ts                  |   98.61 |       96 |     100 |   98.39 | 113
  number-generator-service.ts      |   93.18 |       75 |    87.5 |    92.5 | 39-41
 src/domain/health-check           |     100 |      100 |     100 |     100 |
  health-check-controller.ts       |     100 |      100 |     100 |     100 |
  health-check-service.ts          |     100 |      100 |     100 |     100 |
 src/domain/user                   |   99.09 |    95.83 |     100 |   98.92 |
  auth-service.ts                  |     100 |      100 |     100 |     100 |
  user-controller.ts               |     100 |      100 |     100 |     100 |
  user-params-validator-service.ts |     100 |      100 |     100 |     100 |
  user-service.ts                  |   97.73 |    91.67 |     100 |    97.3 | 65
 src/infra/api-docs                |     100 |      100 |     100 |     100 |
  swagger-config.ts                |     100 |      100 |     100 |     100 |
 src/infra/graphql                 |   36.36 |        0 |       0 |   51.28 |
  Mutation.ts                      |    12.5 |      100 |       0 |   22.22 | 13-58
  Query.ts                         |   41.67 |        0 |       0 |      50 | 9-27
  apollo-server-config.ts          |     100 |      100 |     100 |     100 |
 src/infra/logger                  |   90.63 |    66.67 |     100 |   89.29 |
  winston-config-stream.ts         |   90.63 |    66.67 |     100 |   89.29 | 19-21
 src/infra/repository              |   65.02 |    39.13 |   72.97 |   67.03 |
  in-memory-repository.ts          |   99.07 |    89.29 |     100 |   99.02 | 188
  mongo-repository.ts              |   23.91 |     4.88 |   16.67 |   25.93 | ...,41-46,50-60,65-80,84-95,99-107,111-118,130-149
  respository.interface.ts         |     100 |      100 |     100 |     100 |
 src/infra/repository/mongo/models |   55.56 |      100 |       0 |   55.56 |
  game.ts                          |   55.56 |      100 |       0 |   55.56 | 64-66,73
  user.ts                          |   55.56 |      100 |       0 |   55.56 | 32-34,41
-----------------------------------|---------|----------|---------|---------|----------------------------------------------------

Test Suites: 7 passed, 7 total
Tests:       1 skipped, 54 passed, 55 total
Snapshots:   0 total
Time:        32.475 s
Ran all test suites.
```

### Continuous Integration

This repository uses a CircleUI workflow that runs all tests on every commit. Contributors can see the results of the test before their merge their PR.

![circle_ci_pic](/src/assets/circle_ci.png)

### Logging

This app uses Winston and Morgan middleware to log activity to stdout and to a local file.

The local file is rotated everyday.

<img src="/src/assets/logs.png" alt="drawing" width="350"/>

## Authors

Created by:

- [Tu Vo](https://github.com/tuvo1106)
- Email: tuvo1106@gmail.com


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

### [0.6.0] - 2021-05-13

#### Added

- Difficulty levels now determine number of positions on board
  - i.e. `easy` only has 3 numbers on board

#### Changed

- Refactor `getPlayerGuess` to dynamically get number of guesses
  - Fix bug that users can enter in a blank string
  - Remove unclear validator method for floats, use `Number.isInteger` instead
- Enhance tests in gameService to have better descriptions
  - Also add additional tests to cover more cases
- API no longer passes back user password in any request

### [0.5.2] - 2021-05-06

#### Added

- Security middleware:
  - express-rate-limiter - protection against DDOS attacks
  - express-mongo-sanitize - protection against SQL injections
  - hpp - protection against HTTP parameter pollution
  - helmet - secures app by setting various HTTP headers

#### Changed

- Encrypt user passwords before persisting them
- Simplify seed script

### [0.5.1] - 2021-05-06

#### Added

- Test to cover edge case
- Function to set default env variables if they are not set

#### Changed

- Fix bug on user deletion
- Update README.md

### [0.5.0] - 2021-05-05

#### Added

- Dockerfiles for MongoDB/Node containers
- MongoRepository implementation
- Jest setup script for different environments
- Game and User mongoose schemas

#### Changed

- Change ID size to be 24 characters to match MongoDB ID's
- Change seed script to be database agnostic
- Change tests to be database agnostic

### [0.4.0] - 2021-05-04

#### Added

- GraphQL support for all endpoints
- Custom validator classes instead of relying on outside lib
  - Allows for better decoupling of application logic
- UserParamsValidatorService
- More tests
- More logging at persistence layer

#### Changed

- Update Swagger docs
- Remove validation middleware
- Trim controllers to have little to none business logic

### [0.3.2] - 2021-05-03

#### Added

- Swagger API documentation

#### Changed

- Update README.md

### [0.3.1] - 2021-05-03

#### Added

- NumberGeneratorService tests
- More tests for better test coverage

#### Changed

- Update README.md
- Change max number in game board to 7

### [0.3.0] - 2021-05-03

#### Added

- Password field on user
- More tests for better test coverage

#### Changed

- Game will allow users to sign in
- Remove logging from game CLI
- UserFactory can now set ids and passwords
- Update tests to ensure error messages are correct

### [0.2.0] - 2021-05-03

#### Added

- Game CLI
- Game, Server, Console, Inquirer classes

#### Changed

- Change Score object to also contain previous guesses
- Update winston to utilize chalk colors

### [0.1.0] - 2021-05-02

#### Added

- GameParamsValidatorService
- Script to seed database in 'development` env

#### Changed

- Update logic in GameController/GameService to process guesses
- Change NumberGeneratorService to be a singleton class

### [0.0.3] - 2021-05-02

#### Added

- Environmental variables
- Game controller/service and tests
- GameStatus enum
- GameFactory
- GameEntity
- NumberGeneratorService

#### Changed

- Rename status enum
- Change validateRequest behavior to log out errors
- Update Repository abstract class and its implementations

### [0.0.2] - 2021-04-30

#### Added

- User controller/service and tests
- Abstract Repository
- InMemoryRepository implementation
- BadRequestError and RequestValidationError
- UserFactory
- AppHealth interface
- ValidateRequest middleware
- UserEntity
- Validation middleware (express-validator)

#### Changed

- Renamed various interface files
- Script names

### [0.0.1] - 2021-04-29

#### Added

- Express app
- Logging middleware (Winston, Morgan)
- Health Check controller/service
- Jest setup
- Abstract Custom Error
- NotFoundError
- Error handling middleware
- Utility function (createDirIfNotExists)
- CHANGELOG and NOTES
- CircleCI pipeline


src/index.ts
import app from './app'
import mongoose from 'mongoose'

import { seedDatabase } from './application/scripts/seed-database'
import { logger } from './infra/logger/winston-config-stream'

const startApp = () => {
  const PORT = process.env.PORT || 3000
  checkEnvVariables()
  app.listen(PORT, async () => {
    if (process.env.DB === 'mongo') {
      await connectToMongo()
    }
    logger.info(`Connected to Express on port: ${PORT}.`)
    seedDatabase()
  })
}

const checkEnvVariables = () => {
  if (process.env.DB !== 'local' && process.env.DB !== 'mongo') {
    logger.warn(
      'The following env variable is not set: `DB`. Defaulting to `local`.'
    )
    process.env.DB = 'local'
  }
  if (process.env.NODE_ENV === undefined || process.env.null) {
    logger.warn(
      'The following env variable is not set: `NODE_ENV`. Defaulting to `development`.'
    )
    process.env.NODE_ENV = 'development'
  }
}

const connectToMongo = async () => {
  const MONGO_PORT = process.env.MONGO_PORT || 27017
  try {
    await mongoose.connect(`mongodb://mongo:${MONGO_PORT}/mastermind_v2`, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    })
    logger.info(`Connected to MongoDB on port: ${MONGO_PORT}.`)
  } catch (err) {
    logger.error(err)
    logger.error('Could not connect to MongoDB. Using local database instead.')
    process.env.DB = 'local'
  }
}

startApp()
